sw/device/lib/dif/dif_otp_ctrl.h

```diff
@@ -0,0 +1,773 @@
+    dif_otp_ctrl_toggle_t *state);
+
+/**
+ * Sets whether a particular interrupt is currently enabled or disabled.
+ *
+ * @param otp An OTP handle.
+ * @param irq An interrupt type.
+ * @param state The new toggle state for the interrupt.
+ * @return The result of the operation.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_otp_ctrl_result_t dif_otp_ctrl_irq_set_enabled(const dif_otp_ctrl_t *otp,
+                                                   dif_otp_ctrl_irq_t irq,
+                                                   dif_otp_ctrl_toggle_t state);
+
+/**
+ * Forces a particular interrupt, causing it to be serviced as if hardware had
+ * asserted it.
+ *
+ * @param otp An OTP handle.
+ * @param irq An interrupt type.
+ * @return The result of the operation.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_otp_ctrl_result_t dif_otp_ctrl_irq_force(const dif_otp_ctrl_t *otp,
+                                             dif_otp_ctrl_irq_t irq);
+
+/**
+ * Disables all interrupts, optionally snapshotting all toggle state for later
+ * restoration.
+ *
+ * @param otp An OTP handle.
+ * @param[out] snapshot Out-param for the snapshot; may be `NULL`.
+ * @return The result of the operation.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_otp_ctrl_result_t dif_otp_ctrl_irq_disable_all(
+    const dif_otp_ctrl_t *otp, dif_otp_ctrl_irq_snapshot_t *snapshot);
+
+/**
+ * Restores interrupts from the given snapshot.
+ *
+ * This function can be used with `dif_otp_ctrl_irq_disable_all()` to temporary
+ * interrupt save-and-restore.
+ *
+ * @param otp An OTP handle.
+ * @param snapshot A snapshot to restore from.
+ * @return The result of the operation.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_otp_ctrl_result_t dif_otp_ctrl_irq_restore_all(
+    const dif_otp_ctrl_t *otp, const dif_otp_ctrl_irq_snapshot_t *snapshot);
+
+/**
+ * Checks whether the OTP controller has a particular status bit set.
+ *
+ * @param otp An OTP handle.
+ * @param status The status to check for.
+ * @param[out] has_status Out-param for the status bit.
+ * @return The result of the operation.
+ */
+// Open Q: Do we want a "get all" that returns the whole status register for
+// people to pull values out of themselves with bitfield.h?
+DIF_WARN_UNUSED_RESULT
+dif_otp_ctrl_result_t dif_otp_ctrl_has_status(const dif_otp_ctrl_t *otp,
+                                              dif_otp_ctrl_status_t status,
+                                              bool *has_status);
+
+/**
+ * Looks up the root cause associated with a particular controller status.
+ *
+ * This function should only be called with values of `status` that correspond
+ * to an error condition.
+ *
+ * @param otp An OTP handle.
+ * @param status The error status to root-cause.
+ * @param[out] error Out-param for the root cause.
+ * @return The result of the operation.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_otp_ctrl_result_t dif_otp_ctrl_root_cause_error(
+    const dif_otp_ctrl_t *otp, dif_otp_ctrl_status_t status,
+    dif_otp_ctrl_error_t *error);
```

Comment:
How does this sound?
```
struct {
  dif_otp_ctrl_status_vec_t statuses;
  dif_otp_ctrl_error_t errors[kErrors];
};
```

Reply:
Sounds good, I think we forgot that `dif_otp_ctrl_status_vec_t` could contain more than one status, so of course the errors has to be an array rather than a singular value.

Reply:
Unfortunately, this means that status_t *has* to be bit indices, and not bit masks. =/

Reply:
Ah, so it does.